dotcom-rendering/src/lib/getFrontsBannerAdPositions.ts

```diff
@@ -0,0 +1,224 @@
+			if (groupedCounts.standard === 9) {
+				return 3;
+			} else if (
+				groupedCounts.standard === 5 ||
+				groupedCounts.standard === 6 ||
+				groupedCounts.standard === 7 ||
+				groupedCounts.standard === 8
+			) {
+				return 2.5;
+			} else if (
+				groupedCounts.standard === 1 ||
+				groupedCounts.standard === 3 ||
+				groupedCounts.standard === 4
+			) {
+				return 1.5;
+			} else if (groupedCounts.standard === 2) {
+				return 1;
+			}
+			return 1;
+
+		default:
+			return 1; // Unknown collection type.
+	}
+};
+
+const canAdGoBelowCollection = (
+	heightSinceAd: number,
+	containerPalette: DCRCollectionType['containerPalette'],
+) => {
+	if (containerPalette === 'Branded') {
+		return false;
+	}
+
+	return heightSinceAd >= 3;
+};
+
+const canAdGoAboveNextCollection = (
+	heightSinceAd: number,
+	pageId: string,
+	collection: Pick<
+		DCRCollectionType,
+		'displayName' | 'collectionType' | 'containerPalette'
+	>,
+) => {
+	const excludedCollection = frontsBannerExcludedCollections[pageId] ?? [];
+
+	// There isn't a "next collection" if this is the last collection.
+	// Don't insert an ad below this collection if the next collection prohibits ads above it.
+	if (
+		excludedCollection.includes(collection.displayName) ||
+		collection.containerPalette === 'Branded'
+	) {
+		return false;
+	}
+
+	return heightSinceAd >= 3;
+};
+
+/**
+ * Checks if we have reached a condition that indicates that we no longer
+ * want to insert any more fronts-banner ads into the page.
+ */
+const canStillInsertAds = (
+	adPositions: number[],
+	collectionsLength: number,
+	index: number,
+) => {
+	return (
+		adPositions.length < MAX_FRONTS_BANNER_ADS &&
+		// Don't insert ad below the second-last collection (above the last collection)
+		// We serve a merchandising slot below the last collection and we don't want
+		// to sandwich the last collection between two full-width ads.
+		index < collectionsLength - 2
+	);
+};
+
+/**
+ * Decides where ads should be inserted on standard fronts pages.
+ *
+ * Iterates through the collections and decides where fronts-banner
+ * ad slots should be inserted based on the collection properties.
+ */
+const getFrontsBannerAdPositions = (
+	collections: Pick<
+		DCRCollectionType,
+		'displayName' | 'collectionType' | 'containerPalette' | 'grouped'
+	>[],
+	pageId: string,
+): number[] => {
+	let heightSinceAd = 0;
+	const adPositions = [];
+
+	for (const [index, collection] of collections.entries()) {
+		if (!canStillInsertAds(adPositions, collections.length, index)) {
+			break;
+		}
+
+		const nextCollection = collections[index + 1];
+		if (!nextCollection) {
+			break;
+		}
+
+		heightSinceAd += getCollectionHeight(collection);
+
+		if (
+			canAdGoBelowCollection(
+				heightSinceAd,
+				collection.containerPalette,
+			) &&
+			canAdGoAboveNextCollection(heightSinceAd, pageId, nextCollection)
+		) {
+			adPositions.push(index + 1);
+			heightSinceAd = 0;
+		}
+	}
+
+	return adPositions;
+};
```

Comment:
Not a blocker: this is really a reducer, isn’t it? Could we make it look like one?